test/oncluster/scenario_github_test.go

```diff
@@ -0,0 +1,90 @@
+//go:build oncluster
+// +build oncluster
+
+package oncluster
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"gotest.tools/v3/assert"
+	common "knative.dev/func/test/common"
+)
+
+/*
+Test scenario covered here:
+ - As a Developer I want my function stored on a public GitHub repo to get deployed on my cluster
+
+Notes:
+ * The function used as input for this scenario is stored in this repository at /test/oncluster/testdata/simplefunc
+
+ * On a CI Pull Request action (env CI="true") the branch used on the on-cluster test is the pull request reference.
+   The equivalent deploy func command would look like this:
+
+   func deploy --remote \
+     --git-url https://github.com/knative/func \
+     --git-dir test/oncluster/testdata/simplefunc \
+     --git-branch refs/pull/1650/head
+
+ * When not on CI,  the branch used will be "main" and the repository is https://github.com/knative/func.git
+
+*/
+
+func resolveGitVars() (gitRepoUrl string, gitRef string) {
+	var githubRepository = "knative/func"
+	gitRef = "main"
+	if ci, _ := os.LookupEnv("CI"); ci == "true" {
+		// On a GitHub Action (Pull Request)
+		// https://docs.github.com/en/actions/learn-github-actions/variables
+		githubRepository, _ = os.LookupEnv("GITHUB_REPOSITORY")
+		gitRef, _ = os.LookupEnv("GITHUB_REF")
+	}
+	gitRepoUrl = "https://github.com/" + githubRepository + ".git"
+	return
+}
+
+// TestGitHubFunc tests the following use case:
+//   - As a Developer I want my function stored on a public GitHub repo to get deployed on my cluster
+func TestGitHubFunc(t *testing.T) {
+
+	var tmpDir = t.TempDir()
+	var tmpRepo = "knative-func"
+	var cloneDir = filepath.Join(tmpDir, tmpRepo)
+
+	var githubRepo, githubRef = resolveGitVars()
+	var funcName = "simplefunc"
+	var funcContextDir = filepath.Join("test", "oncluster", "testdata", funcName)
+	var funcPath = filepath.Join(cloneDir, funcContextDir)
+
+	// -- Clone Func from GITHUB and checkout Branch
+	sh := common.NewShellCmd(t, tmpDir)
+	sh.ShouldFailOnError = true
+	sh.ShouldDumpOnSuccess = false
+	sh.Exec("git init " + tmpRepo)
+
+	sh.SourceDir = cloneDir
+	sh.Exec("git remote add origin " + githubRepo)
+	sh.Exec("git fetch --recurse-submodules=yes --depth=1 origin --update-head-ok --force " + githubRef)
+	sh.Exec("git checkout FETCH_HEAD")
```

Comment:
Why not just `git clone`?

Reply:
I found this approach than clone consider the side the repo is getting... (I bother this git approach from the git-clone taks itself).. it does not clone the entire git tree.. just a shallow copy of the repo. Here we just need access to the `func.yaml` locally. TBH I think `func deploy --remote --git-url ...` should not even require a func locally ...